packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -730,14 +750,27 @@ Future<DateTime> showDatePicker({
   @required BuildContext context,
   @required DateTime initialDate,
   @required DateTime firstDate,
-  @required DateTime lastDate
+  @required DateTime lastDate,
+  SelectableDayPredicate selectableDayPredicate
 }) async {
+  if (initialDate.isBefore(firstDate))
+    return new Future<DateTime>.error(
+       new ArgumentError('initialDate must be on or after firstDate'));
```

Comment:
This function is already `async`, so you can just throw an exception and it will cause the future to complete with an error.

Usually for these sort of argument validations, we just use `assert`.  If we want to provide a detailed error message, then we'll throw a `FlutterError`, but that doesn't need necessary in this case.